File: services/chess_service/src/main/java/net/chess_platform/chess_service/chess/move/CaptureMove.java
package net.chess_platform.chess_service.chess.move;

import net.chess_platform.chess_service.chess.Chessboard;
import net.chess_platform.chess_service.chess.piece.Piece;
import net.chess_platform.chess_service.chess.piece.Piece.Color;

public class CaptureMove extends AbstractMove {

    private Piece capturedPieceInstance;

    public CaptureMove(Piece.Type piece, Color color, Position from, Position to) {
        super(piece, color, from, to, Type.CAPTURE);
    }

    @Override
    public void execute(Chessboard board) {
        super.execute(board);
        var from = getFrom();
        var to = getTo();

        capturedPieceInstance = board.getPiece(to);

        board.setPiece(getMovedPieceInstance(), to);
        board.setPiece(null, from);
    }

    @Override
    public void undo(Chessboard board) {
        var from = getFrom();
        var to = getTo();

        board.setPiece(getMovedPieceInstance(), from);
        board.setPiece(capturedPieceInstance, to);

        capturedPieceInstance = null;

        super.undo(board);
    }
}

File: app/src/chess/moves/CaptureMove.ts
import { toAlgebraicNotation, toCodePoint } from "../chess-view";
import { Chessboard } from "../Chessboard";
import { Color } from "../pieces/AbstractPiece";
import { Piece, PieceType } from "../pieces/Piece";
import { AbstractMove } from "./AbstractMove";
import { CheckStatus, Position } from './Move';

export class CaptureMove extends AbstractMove {

    private capturedPiece?: Piece;

    constructor(piece: PieceType | null, color: Color, from: Position, to: Position, checkStatus?: CheckStatus, timestamp?: number) {
        super(piece, color, from, to, "CAPTURE", checkStatus, timestamp);
    }

    override execute(board: Chessboard): void {
        super.execute(board);

        const from = this.getFrom();
        const to = this.getTo();
        this.capturedPiece = board.getPiece(to.row, to.col)!;

        board.setPiece(to.row, to.col, this.getMovedPieceInstance()!);
        board.setPiece(from.row, from.col, null);
    }

    override undo(board: Chessboard): void {
        const from = this.getFrom();
        const to = this.getTo();

        board.setPiece(from.row, from.col, this.getMovedPieceInstance()!);
        board.setPiece(to.row, to.col, this.capturedPiece!);
        this.capturedPiece = undefined;

        super.undo(board);
    }

    toAlgebraicNotationImpl(): string {
        return `${toCodePoint(this.getColor() + "-" + this.getPiece())}${toAlgebraicNotation(this.getFrom())}x${toAlgebraicNotation(this.getTo())}`;
    }
}